components/raftstore/src/store/fsm/peer.rs

```diff
@@ -6555,6 +6561,58 @@ where
     fn register_report_region_buckets_tick(&mut self) {
         self.schedule_tick(PeerTick::ReportBuckets)
     }
+
+    /// Check whether the peer is pending on applying raft logs.
+    ///
+    /// If pending, the peer will be recorded, until the pending logs are
+    /// applied. And after it completes applying, it will be removed from
+    /// the recording list.
+    fn on_check_peer_complete_apply_logs(&mut self) {
+        if self.fsm.peer.pending_on_apply.is_none() {
+            return;
+        }
+
+        let peer_id = self.fsm.peer.peer_id();
+        let applied_idx = self.fsm.peer.get_store().applied_index();
+        let last_idx = self.fsm.peer.get_store().last_index();
+        // If the peer is newly added or created, no need to check the apply status.
+        if last_idx <= RAFT_INIT_LOG_INDEX {
+            self.fsm.peer.pending_on_apply = None;
+            return;
+        }
+        assert!(self.fsm.peer.pending_on_apply.is_some());
+        // If the peer has large unapplied logs, this peer should be recorded until
+        // the lag is less than the given threshold.
+        if last_idx >= applied_idx + self.ctx.cfg.leader_transfer_max_log_lag {
+            if !self.fsm.peer.pending_on_apply.unwrap() {
+                let mut meta = self.ctx.store_meta.lock().unwrap();
+                meta.pending_apply_peers.insert(peer_id);
```

Comment:
Nit: How about stashing busy peers in ctx and later merging them into store_meta in `fn end()`? So we can save acquiring locks.

Reply:
If this method is call for the first 5 minutes after restart, we do need this optimization.

Reply:
No need for it. It's safe and no extra costs, as the `pending_on_apply` will be checked in advance.